octobot_tentacles_manager/uploaders/s3_uploader.py

```diff
@@ -0,0 +1,96 @@
+#  Lesser General Public License for more details.
+#
+#  You should have received a copy of the GNU Lesser General Public
+#  License along with this library.
+import os
+import aioboto3
+
+import octobot_tentacles_manager.uploaders.uploader as uploader
+
+
+class S3Uploader(uploader.Uploader):
+    ENV_S3_API_KEY = "S3_API_KEY"
+    ENV_S3_API_SECRET_KEY = "S3_API_SECRET_KEY"
+    ENV_S3_ENDPOINT_URL = "S3_ENDPOINT_URL"
+    ENV_S3_BUCKET_NAME = "S3_BUCKET_NAME"
+    ENV_S3_REGION_NAME = "S3_REGION_NAME"
+
+    S3_BOTO_KEY = "s3"
+
+    def __init__(self):
+        super().__init__()
+        self.s3_session: aioboto3.Session = aioboto3.Session()
+        self.s3_api_key: str = os.getenv(S3Uploader.ENV_S3_API_KEY, None)
+        self.s3_api_secret_key: str = os.getenv(S3Uploader.ENV_S3_API_SECRET_KEY, None)
+        self.s3_endpoint_url: str = os.getenv(S3Uploader.ENV_S3_ENDPOINT_URL, None)
+        self.s3_bucket_name: str = os.getenv(S3Uploader.ENV_S3_BUCKET_NAME, None)
+        self.s3_region_name: str = os.getenv(S3Uploader.ENV_S3_REGION_NAME, None)
+        if None in [self.s3_api_key, self.s3_api_secret_key, self.s3_endpoint_url,
+                    self.s3_bucket_name, self.s3_region_name]:
+            raise TypeError("Some s3 environment variables are missing, please ensure that "
+                            "S3_API_KEY, S3_API_SECRET_KEY, S3_BUCKET_NAME, S3_REGION_NAME "
+                            "and S3_ENDPOINT_URL are defined.")
+
+    async def upload_file(self, upload_path: str, file_path: str, destination_file_name: str = None) -> int:
+        """
+        Upload file in s3
+        :param upload_path: the upload path
+        :param file_path: the file local path
+        :param destination_file_name: the file name in s3 bucket (optional : default file_path basename)
+        :return: the result of _upload
+        """
+        dest_file_name: str = destination_file_name \
+            if destination_file_name is not None else os.path.basename(file_path)
+        upload_file_url: str = f"{upload_path}{dest_file_name}"
+        self.logger.info(f"Uploading {file_path} to nexus at {upload_file_url}...")
+        return await self._upload(object_name=upload_file_url, local_file_path=file_path)
+
+    async def upload_folder(self, upload_path: str, folder_path: str, destination_folder_name: str = None) -> int:
+        """
+        Upload folder content in s3
+        :param upload_path: the upload path
+        :param folder_path: the folder local path
+        :param destination_folder_name: the folder name in s3 bucket (optional : default folder_path basename)
+        :return: the sum of all of _upload returns
+        """
+        error_count: int = 0
+        dest_folder_name: str = destination_folder_name \
+            if destination_folder_name is not None else os.path.basename(folder_path)
+        upload_folder_url: str = f"{upload_path}{dest_folder_name}"
+        for file_path in os.listdir(folder_path):
```

Comment:
should this be recursive ?

Reply:
It's doing the job (see tests). Do you think I'm missing something ?

Reply:
Right, done and tested in test_s3_uploader.py